python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+    execution_timeout: Optional[timedelta] = None,
+    run_timeout: Optional[timedelta] = None,
+    task_timeout: Optional[timedelta] = None,
+    id_reuse_policy: temporalio.common.WorkflowIDReusePolicy = temporalio.common.WorkflowIDReusePolicy.ALLOW_DUPLICATE,
+    retry_policy: Optional[temporalio.common.RetryPolicy] = None,
+    cron_schedule: str = "",
+    memo: Optional[Mapping[str, Any]] = None,
+    search_attributes: Optional[Mapping[str, Any]] = None,
+    header: Optional[Mapping[str, Any]] = None,
+) -> ChildWorkflowHandle[Any, Any]:
+    # Do stuff
+```
+
+Notes:
+
+* `ChildWorkflowHandle.signal` will have the same typed overloads that `WorkflowHandle.signal` does
+* `ChildWorkflowHandle.cancel` is present to simplify cancellation. See "temporalio.workflow Scoped Cancellation" for
+  more discussion.
+* `start_child` was considered instead of `start_child_workflow`, but this seems clearer. Can be convinced otherwise.
+
+#### `temporalio.workflow.execute_child_workflow()`
+
+Shortcut for `temporalio.workflow.start_child_workflow()` + `handle.result()`
+
+#### `temporalio.workflow.ChildWorkflowConfig`
+
+`TypedDict` of `start_child_workflow`/`execute_child_workflow` configuration options.
+
+#### `temporalio.workflow.continue_as_new()`
+
+```python
+# Overload for self (cannot type args)
+@overload
+def continue_as_new(
+    *args: Any,
+    # ... (excludes "workflow" kwarg)
+) -> NoReturn:
+    pass
+
+# Overload for no-param workflow
+# TODO(cretz): Fix when https://github.com/python/mypy/issues/11833 fixed
+@overload
+def continue_as_new(
+    *,
+    workflow: Callable[[WorkflowType], Any],
+    # ...
+) -> NoReturn:
+    pass
+
+# Overload for single-param workflow
+# TODO(cretz): Fix when https://github.com/python/mypy/issues/11833 fixed
+@overload
+def continue_as_new(
+    arg: LocalParamType,
+    /,
+    *,
+    workflow: Callable[[WorkflowType, LocalParamType], Any],
+    # ...
+) -> NoReturn:
+    pass
+
+# Overload for string workflow
+@overload
+def continue_as_new(
+    *args: Any,
+    workflow: str,
+    # ...
+) -> NoReturn:
+    pass
+
+def continue_as_new(
+    *args: Any,
+    workflow: Union[None, Callable, str] = None,
+    task_queue: Optional[str] = None,
+    memo: Optional[Mapping[str, Any]] = None,
+    search_attributes: Optional[Mapping[str, Any]] = None,
+    run_timeout: Optional[timedelta] = None,
+    task_timeout: Optional[timedelta] = None,
+) -> NoReturn:
+    # Do stuff
+```
+
+Notes:
+
+* Like all other SDK languages so far, we can't reasonably provide types for args on the "self" continue as new
+* More investigation is needed on whether I can stop the event loop inside the continue as new without it returning
+  * I probably can't so I probably need to make this call `async`, but I probably don't have to throw
+  * While I thought I could just stop the event loop and then return, that doesn't prevent the next line of code from
```

Comment:
That should be fine as long as it's deterministic

Reply:
Even if it is deterministic, I want to stop execution immediately on continue-as-new which I believe requires it being async. More research is required here.

Reply:
👍